protoc-gen-twirp/generator.go

```diff
@@ -1018,7 +1028,8 @@ func (t *twirp) generateServerJSONMethod(service *descriptor.ServiceDescriptorPr
 	t.P(`  resp.Header().Set("Content-Type", "application/json")`)
 	t.P(`  resp.WriteHeader(`, t.pkgs["http"], `.StatusOK)`)
 	t.P(`  if _, err = resp.Write(buf.Bytes()); err != nil {`)
-	t.P(`    `, t.pkgs["log"], `.Printf("errored while writing response to client, but already sent response status code to 200: %s", err)`)
+	t.P(`    err = wrapErr(err, "failed to write response to client")`)
+	t.P(`    callError(ctx, s.hooks,`, t.pkgs["twirp"], `.InternalErrorWith(err))`)
```

Comment:
Using an "Internal" error to call the hook seems weird. Write errors here are caused by things like clients that have reset the connection .. that seems like an "external" error :)

I read through the error codes we have and didn't see one that fits well.

If we know that it's because the client disconnected, it might be appropriate to return a "Canceled" error .. but if we do that, then maybe we should check against the Context value (which net/http should have canceled)? Is that complexity worthwhile?

We could make a rough attempt to classify the error into "Canceled", maybe "DeadlineExceeded", and maybe others .. and use "Unknown" for anything we don't understand?

Reply:
Yeah, "Unknown" might be more appropriate here for now. I don't think I want to get into the weeds of diagnosing whether it was a context cancellation. That seems hard to get right, and I'd rather conservatively label this as unknown.

I've been also mulling a `"TransportFailure"` error code, to describe all these sorts of low-level problems. I don't think this is the time to add it, but this might be a good use for it.